README.md
WoWChat -- README
=================

WoWChat is a Discord integration chat bot for old versions of World of Warcraft.

**It does NOT support WoW Classic or Retail servers.**

New Requirement: Due to Discord changes, you must check PRESENCE INTENT and SERVER MEMBERS INTENT under Privileged Gateway Intents in Discord developer portal.

Currently supported versions are:
  * Vanilla
  * The Burning Crusade
  * Wrath of the Lich King
  * Cataclysm (4.3.4 build 15595)
  * Mists of Pandaria (5.4.8 build 18414)

Features:
* Clientless (Does not need the WoW Client to be open to run)
* Seamless Chat integration of channels between WoW and Discord
  * Guild chat, Officer chat, Local chat, Emotes, Custom Channels.
  * In-game links (items, spells, ...) are displayed as links to classicdb or twinstar, depending on the expansion
  * Configurable message format
* Smart Tagging
  * Tag players on Discord from WoW using @and_part_or_all_of_their_name.
  * You can also tag @here and @everyone and "@Even Roles With Spaces" (include quotes around them).
* Custom commands
  * Check who is online in your guild
  * Query other players in the world
* Runs as a Java program, and therefore works on Windows, Mac, and Linux.

## How it works
The bot uses Discord's API to login to your Discord server. It then uses supplied information
to login as a WoW character onto your chosen server. Once it logs in to WoW as a character,
and sees the configured channels. It will relay messages in Discord chat and WoW chat respectively.

##### DO NOT, under any circumstances, use this bot on an account with existing characters!
Even though this bot does not do anything malicious, some servers may not like a bot connecting, and GMs may ban the account!

##### Watching GD's WotLK chat:
![gd-echoes](https://raw.githubusercontent.com/fjaros/wowchat/master/images/example1.png)

##### Talking in Guild Chat:
![guild-chat-construct](https://raw.githubusercontent.com/fjaros/wowchat/master/images/example2.png)

## Setup & Prerequisites

1. First you will want to create a Discord Bot on your discord account:
   * Go to https://discordapp.com/developers/applications/
   * Sign into your Discord account if necessary and click "Create an application"
   * Change the application name to something meaningful like "WoW Chat"
   * On the left click the Bot tab
   * Add a Bot
   * Uncheck Public Bot option
   * **Check PRESENCE INTENT and SERVER MEMBERS INTENT under "Privileged Gateway Intents"** This is important! Without it, your bot will not work!
   * Under token click Copy. This is the value WoW Chat will use to login to Discord.
2. Configure WoW Chat by opening `wowchat.conf` in a text editor.
   * You can also create your own file, using the supplied `wowchat.conf` as a template.
   * In section **discord**:
     * **token**: Paste the above copied Bot token, or set the DISCORD_TOKEN environment variable.
     * **enable_dot_commands**: If set to 1, it will not format outgoing messages starting with ".", enabling you to send things like ".s in" to the server directly. If set to 0, it will format these messages like regular messages.
     * **dot_commands_whitelist**: If empty, it will allow or disallow dot commands based on **enable_dot_commands** setting. If any command is listed here, the bot will ONLY allow those specific comamnds to be sent in game.
     * **enable_commands_channels**: A list of channels for which to allow commands. If not specified or empty, the bot will allow commands from all channels.
   * In section **wow**:
     * **platform**: Leave as **Mac** unless your target server has Warden (anticheat) disabled AND it is blocking/has disabled Mac logins. In this case put **Windows**.
     * **locale**: Optionally specify a locale if you want to join locale-specific global channels. **enUS** is the default locale.
     * **enable_server_motd**: **0** to ignore sending server's MotD. **1** to send server's MotD as a SYSTEM message.
     * **version**: put either 1.12.1, 2.4.3, 3.3.5, 4.3.4, or 5.4.8 based on the server's expansion.
     * **build**: you can include a build=<build number> setting in the config, if you are using a custom build version on your server.
     * **realmlist**: this is server's realmlist, same as in your realmlist.wtf file.
     Example values are logon.lightshope.org or wow.gamer-district.org
     * **realm**: This is the realm name the Bot will connect to.
     It is the Text shown on top of character list window. Put ONLY the name, do NOT put the realm type like PVP or PVE.
     In the following example, the **realm** value is The Construct
     * ![realm-construct](https://raw.githubusercontent.com/fjaros/wowchat/master/images/example3.png)
     * **account**: The bot's WoW game account, or set the WOW_ACCOUNT environment variable.
     * **password**: The bot's WoW game account password, or set the WOW_PASSWORD environment variable.
     * **character**: Your character's name as would be shown in the character list, or set the WOW_CHARACTER environment variable.
   * In section **guild**:
     * This section sets up guild notifications on Discord.
     * For each notification, **online**, **offline**, **joined**, **left**, **motd**, **achievement** specify:
       * **enabled**: **0** to not display in Discord, **1** to display in Discord
       * **format**: How to display the message.
       * **channel**: Optional channel **name** OR **ID** where to display message instead of the default guild chat channel.
   * In section **chat**:
     * This section sets up the channel relays between Discord and WoW. You can have an unlimited number of channel relays.
     * **direction**: How do you want to relay each channel, put either
     **wow_to_discord**, **discord_to_wow**, or **both**.
     * **wow** section:
       * In type put one of, **Say**, **Guild**, **Officer**, **Emote**, **Yell**, **System**, **Whisper**, **Channel**. This is the type of chat the Bot will read for this section.
         * If you put **type=Channel**, you also must provide a **channel=name of channel** value.
       * In format put how you want to display the message, supported replacable values are **%time**, **%user**, **%message**, and **%channel** if above type is **Channel**.
       * **filters**: See filters section. If a channel configuration has this section, it will override the global filters and use these instead for this channel.
         * If this is in the **wow** section, it will filter Discord->WoW messages.
       * Optionally in **id**, specify the channel ID if your server has a non-standard global channel.
     * **discord** section:
       * **channel**: The discord channel **name** OR **ID** where to display the message. **It is advised to use channel ID here instead of name, so the bot does not stop working when the channel name is changed.**
         * To see channels' IDs, you must enable Developer mode in Discord under User Settings -> Appearance -> Advanced.
       * **format**: Same options as in **wow** section above.
       * **filters**: See filters section. If a channel configuration has this section, it will override the global filters and use these instead for this channel.
         * If this is in the **discord** section, it will filter WoW->Discord messages.
   * In section **filters**:
     * This section specifies filters for chat messages to be ignored by the bot. It works for both directions, Discord to WoW and WoW to Discord. It can be overriden in each specific channel configuration as stated above.
     * **enabled**: **0** to globally disable all filters, **1** to enable them.
     * **patterns**: List of Java Regex match patterns. If the incoming messages matches any one of the patterns and filters are enabled, it will be ignored.
       * When ignored, the message will not be relayed; however it will be logged into the bot's command line output prepended with the word FILTERED.

3. Invite your bot to Discord
   * Go back to https://discordapp.com/developers/applications/ and click your new Bot application.
   * In browser enter: https://discordapp.com/oauth2/authorize?client_id=CLIENT_ID&scope=bot
     * Replace **CLIENT_ID** with the value from Discord applications page.
   * Setup the bot with the correct Discord roles/permissions to enter your desired channels.

## Run
1. Download the latest ready-made binary from github releases: https://github.com/fjaros/wowchat/releases
   * **Make sure you have a Java Runtime Environment (JRE) 1.8 or higher installed on your system!**
   * **On Windows**: Edit wowchat.conf as above and run `run.bat`
   * **On Mac/Linux**: Edit wowchat.conf as above and run `run.sh`

OR to compile yourself:
1. WoW Chat is written in Scala and compiles to a Java executable using [maven](https://maven.apache.org).
2. It uses Java JDK 1.8 and Scala 2.12.12.
3. Run `mvn clean package` which will produce a file in the target folder called `wowchat-1.3.8.zip`
4. unzip `wowchat-1.3.8.zip`, edit the configuration file and run `java -jar wowchat.jar <config file>`
   * If no config file is supplied, the bot will try to use `wowchat.conf`


CHANGELOG
2022-02-20 - v1.3.8
+ Updates to used libraries.
+ Added build hashes and support for more 1.x vanilla builds.
* Fixed issue with bot not sending messages longer than 255 characters, since this is unsupported by WoW clients.
  Instead, the bot will split up these messages and send them in multiple chat packets.
* Fixed issue with bot not sending Discord messages which are replies.

2020-10-27 - v1.3.4/v1.3.5/v1.3.6
* Update to new Discord Library. This is a mandatory update. Using an old version will not connect to Discord. (Issue #45)
* Added support for Discord Gateway Intents and appropriate requirement messages to users. (Issue #47)
- Changed filter behavior to modify post processed messages. (Issue #35)

2019-09-29 - v1.3.3
+ Enabled setting account, password, and character variables from environmental variables (Commit by Road-block).
+ Implemented correct channel ID handling for global/localized/zone-specific channels like General and Trade (Issue #16).
+ Added config option for turning off tag matching failed notifications.
* Fixed bug causing failed warden responses on some servers which send warden check requests containing null byte.

2019-06-01 - v1.3.2
+ Improved Discord tag matching heuristics:
  + If multiple matches are found, preferred matches are chosen if the match is exact, or if the match contains the tag as an independent word.
  + If a picked match cannot be found, the bot will send the message as is without resolving the tag.
    In addition to whispering the sender, it will also write the error into the Discord channel, since a lot of servers block
    level 1 characters from whispering.
* Fixed bot breaking on WotLK/Cata when receiving very large responses such as the guild roster for a 900+ member guild.
* Fixed incorrect application of escaping markdown before applying config-based message formatting.

2019-04-24 - v1.3.1
+ Improve who command handling to better return approximate results from a mix of offline guild members and online players.
+ Improve matching of tags in WoW->Discord messages to tag users even if their name is followed by special characters.
+ Enabled escaping of characters in WoW->Discord messages that would format the message on Discord.
* Fixed issue with some servers not sending back character list in case it is requested before anti cheat response.

2019-03-20 - v1.3.0
+ Added support for global and channel-bound chat filters using Java Regex pattern matching.
+ Added whitelist feature to optionally explicitly only allow users using certain dot commands.
+ Added support for restart/shutdown announcement messages to be relayed as System messages.
+ Added command ?gmotd to display your guild's motd.
+ Added option to send guild notifications and guild motd to another Discord channel.
+ Changed behavior of ?who command to always return exact player requested if available on queries returning multiple players.
* Fixed bug in ?who command not working if the server returns a different player count vs the number of players actually displayed.
* Fixed bug in displaying color coded messages if the message contains the $ symbol.
* Fixed rare sync bug on some servers if the server + bot are both running for longer than 24 days.

2019-02-14 - v1.2.10
+ Enabled parsing of attachment (image links) in Discord->WoW messages
+ Added configuration setting for ingoring server MotD.
+ Added support for custom build numbers. Useful for users who have servers that accept modified clients.
+ Added SYSTEM notification for when the bot disconnects from the WoW server.
+ Lowered default reconnect timeout to 10 seconds.
* Fixed promote/demote notifications for MoP.
* Fixed emoji parsing in messages which contain item links.
* Fixed sending of empty messages to WoW for non-text events (users joining Discord, user pinning a message to the channel).
* Fixed case sensitivity in ?who command.

2019-01-14 - v1.2.9
+ Added optional login as Windows platform. This will fix support for Netherwing server while their Mac login is broken.
* Separated player and guild roster into two separate entities. Will fix sometimes random names in ?who command

2019-01-11 - v1.2.8
+ Added support for guild promoted, demoted, and player removed events.
* Fixed bot throwing an error if it received a message for a non-configured Discord channel.
* Fixed bug in logging in with characters using non-ASCII characters.
* Fixed bug in joining channels with non-ASCII characters on Cata/MoP.

2018-12-31 - v1.2.7
+ Added support for looking up Discord channels by ID rather than just name. To see a channel's ID, you must enable
  Developer Mode in Discord under User Settings -> Appearance -> Advanced. Right-clicking the channel will then reveal the ID.
  This is advantageous to use over name, so you need not change bot configuration when a channel name is changed.
* Do not send guild motd notification if motd is empty.
* Fixed logging in on accounts with multiple characters on TBC/WotLK.

2018-12-20 - v1.2.6
+ Added support for more trade links in chat like professions for WotLK/Cata/MoP.
+ Added support for GM sending messages to chats on TBC/WotLK/Cata.
* Fixed failing to relay messages from WoW to Discord on a Discord reconnect.
* Fixed displaying non-latin fonts in Windows command line.

2018-12-07 - v1.2.5
* Fixed inability to connect to servers which "overflow" their port in realm listing.
* Force usage of UTF-8 for the reading of all chat messages. Fixes issues with display non-latin fonts.
* Fixed bot sending multiple messages to WoW per one message in Discord if invited to multiple Discord servers.
+ Added enable_commands_channels config setting. If specified, only enables ?who in the configured channels.

2018-11-28 - v1.2.4
* Fixed initiating 2 reconnect timers if realm server is down.
* Fixed tagging @here to correctly tag @here instead of resolving a name containg "here" first.

src/main/scala/wowchat/common/Config.scala
package wowchat.common

import java.io.File
import java.util

import wowchat.common.ChatDirection.ChatDirection
import wowchat.common.WowExpansion.WowExpansion
import com.typesafe.config.{Config, ConfigFactory}
import wowchat.game.GamePackets

import scala.collection.JavaConverters._
import scala.reflect.runtime.universe.{TypeTag, typeOf}

case class WowChatConfig(discord: DiscordConfig, wow: Wow, guildConfig: GuildConfig, channels: Seq[ChannelConfig], filters: Option[FiltersConfig])
case class DiscordConfig(token: String, enableDotCommands: Boolean, dotCommandsWhitelist: Set[String], enableCommandsChannels: Set[String], enableTagFailedNotifications: Boolean)
case class Wow(locale: String, platform: Platform.Value, build: Option[Int], realmlist: RealmListConfig, account: Array[Byte], password: String, character: String, enableServerMotd: Boolean)
case class RealmListConfig(name: String, host: String, port: Int)
case class GuildConfig(notificationConfigs: Map[String, GuildNotificationConfig])
case class GuildNotificationConfig(enabled: Boolean, format: String, channel: Option[String])
case class ChannelConfig(chatDirection: ChatDirection, wow: WowChannelConfig, discord: DiscordChannelConfig)
case class WowChannelConfig(id: Option[Int], tp: Byte, channel: Option[String] = None, format: String, filters: Option[FiltersConfig])
case class DiscordChannelConfig(channel: String, format: String, filters: Option[FiltersConfig])
case class FiltersConfig(enabled: Boolean, patterns: Seq[String])

object WowChatConfig extends GamePackets {

  private var version: String = _
  private var expansion: WowExpansion = _

  def apply(confFile: String): WowChatConfig = {
    val file = new File(confFile)
    val config = (if (file.exists) {
      ConfigFactory.parseFile(file)
    } else {
      ConfigFactory.load(confFile)
    }).resolve

    val discordConf = config.getConfig("discord")
    val wowConf = config.getConfig("wow")
    val guildConf = getConfigOpt(config, "guild")
    val channelsConf = config.getConfig("chat")
    val filtersConf = getConfigOpt(config, "filters")

    // we gotta load this first to initialize constants that change between versions :OMEGALUL:
    version = getOpt(wowConf, "version").getOrElse("1.12.1")
    expansion = WowExpansion.valueOf(version)

    WowChatConfig(
      DiscordConfig(
        discordConf.getString("token"),
        getOpt[Boolean](discordConf, "enable_dot_commands").getOrElse(true),
        getOpt[util.List[String]](discordConf, "dot_commands_whitelist")
          .getOrElse(new util.ArrayList[String]()).asScala.map(_.toLowerCase).toSet,
        getOpt[util.List[String]](discordConf, "enable_commands_channels")
          .getOrElse(new util.ArrayList[String]()).asScala.map(_.toLowerCase).toSet,
        getOpt[Boolean](discordConf, "enable_tag_failed_notifications").getOrElse(true)
      ),
      Wow(
        getOpt[String](wowConf, "locale").getOrElse("enUS"),
        Platform.valueOf(getOpt[String](wowConf, "platform").getOrElse("Mac")),
        getOpt[Int](wowConf, "build"),
        parseRealmlist(wowConf),
        convertToUpper(wowConf.getString("account")),
        wowConf.getString("password"),
        wowConf.getString("character"),
        getOpt[Boolean](wowConf, "enable_server_motd").getOrElse(true)
      ),
      parseGuildConfig(guildConf),
      parseChannels(channelsConf),
      parseFilters(filtersConf)
    )
  }

  lazy val getVersion = version
  lazy val getExpansion = expansion

  lazy val getBuild: Int = {
    Global.config.wow.build.getOrElse(
      version match {
        case "1.6.1" => 4544
        case "1.6.2" => 4565
        case "1.6.3" => 4620
        case "1.7.1" => 4695
        case "1.8.4" => 4878
        case "1.9.4" => 5086
        case "1.10.2" => 5302
        case "1.11.2" => 5464
        case "1.12.1" => 5875
        case "1.12.2" => 6005
        case "1.12.3" => 6141
        case "2.4.3" => 8606
        case "3.2.2" => 10505
        case "3.3.0" => 11159
        case "3.3.2" => 11403
        case "3.3.3" => 11723
        case "3.3.5" => 12340
        case "4.3.4" => 15595
        case "5.4.8" => 18414
        case _ => throw new IllegalArgumentException(s"Build $version not supported!")
      })
  }

  private def convertToUpper(account: String): Array[Byte] = {
    account.map(c => {
      if (c >= 'a' && c <= 'z') {
        c.toUpper
      } else {
        c
      }
    }).getBytes("UTF-8")
  }

  private def parseRealmlist(wowConf: Config): RealmListConfig = {
    val realmlist = wowConf.getString("realmlist")
    val splt = realmlist.split(":", 2)
    val (host, port) =
      if (splt.length == 1) {
        (splt.head, 3724)
      } else {
        (splt.head, splt(1).toInt)
      }

    RealmListConfig(wowConf.getString("realm"), host, port)
  }

  private def parseGuildConfig(guildConf: Option[Config]): GuildConfig = {
    // make reasonable defaults for old config
    val defaults = Map(
      "promoted" -> (true, "`[%user] has promoted [%target] to [%rank].`"),
      "demoted" -> (true, "`[%user] has demoted [%target] to [%rank].`"),
      "online" -> (false, "`[%user] has come online.`"),
      "offline" -> (false, "`[%user] has gone offline.`"),
      "joined" -> (true, "`[%user] has joined the guild.`"),
      "left" -> (true, "`[%user] has left the guild.`"),
      "removed" -> (true, "`[%target] has been kicked out of the guild by [%user].`"),
      "motd" -> (true, "`Guild Message of the Day: %message`"),
      "achievement" -> (true, "%user has earned the achievement %achievement!")
    )

    guildConf.fold({
      GuildConfig(defaults.mapValues {
        case (enabled, format) => GuildNotificationConfig(enabled, format, None)
      })
    })(guildConf => {
      GuildConfig(
        defaults.keysIterator.map(key => {
          val conf = getConfigOpt(guildConf, key)
          val default = defaults(key)
          key -> conf.fold(GuildNotificationConfig(default._1, default._2, None))(conf => {
            GuildNotificationConfig(
              getOpt[Boolean](conf, "enabled").getOrElse(default._1),
              getOpt[String](conf, "format").getOrElse(default._2),
              getOpt[String](conf, "channel")
            )
          })
        })
          .toMap
      )
    })
  }

  private def parseChannels(channelsConf: Config): Seq[ChannelConfig] = {
    channelsConf.getObjectList("channels").asScala
      .map(_.toConfig)
      .map(channel => {
        val wowChannel = getOpt[String](channel, "wow.channel")

        ChannelConfig(
          ChatDirection.withName(channel.getString("direction")),
          WowChannelConfig(
            getOpt[Int](channel, "wow.id"),
            ChatEvents.parse(channel.getString("wow.type")),
            wowChannel,
            getOpt[String](channel, "wow.format").getOrElse(""),
            parseFilters(getConfigOpt(channel, "wow.filters"))
          ),
          DiscordChannelConfig(
            channel.getString("discord.channel"),
            channel.getString("discord.format"),
            parseFilters(getConfigOpt(channel, "discord.filters"))
          )
        )
    })
  }

  private def parseFilters(filtersConf: Option[Config]): Option[FiltersConfig] = {
    filtersConf.map(config => {
      FiltersConfig(
        getOpt[Boolean](config, "enabled").getOrElse(false),
        getOpt[util.List[String]](config, "patterns").getOrElse(new util.ArrayList[String]()).asScala
      )
    })
  }

  private def getConfigOpt(cfg: Config, path: String): Option[Config] = {
    if (cfg.hasPath(path)) {
      Some(cfg.getConfig(path))
    } else {
      None
    }
  }

  private def getOpt[T : TypeTag](cfg: Config, path: String): Option[T] = {
    if (cfg.hasPath(path)) {
      // evil smiley face :)
      Some(
        (if (typeOf[T] =:= typeOf[Boolean]) {
          cfg.getString(path).toLowerCase match {
            case "true" | "1" | "y" | "yes" => true
            case _ => false
          }
        } else if (typeOf[T] =:= typeOf[String]) {
          cfg.getString(path)
        } else {
          cfg.getAnyRef(path)
        }).asInstanceOf[T]
      )
    } else {
      None
    }
  }
}

object Platform extends Enumeration {
  type Platform = Value
  val Windows, Mac = Value

  def valueOf(platform: String): Platform = {
    platform.toLowerCase match {
      case "win" | "windows" => Windows
      case _ => Mac
    }
  }
}

object WowExpansion extends Enumeration {
  type WowExpansion = Value
  val Vanilla, TBC, WotLK, Cataclysm, MoP = Value

  def valueOf(version: String): WowExpansion = {
    if (version.startsWith("1.")) {
      WowExpansion.Vanilla
    } else if (version.startsWith("2.")) {
      WowExpansion.TBC
    } else if (version.startsWith("3.")) {
      WowExpansion.WotLK
    } else if (version == "4.3.4") {
      WowExpansion.Cataclysm
    } else if (version == "5.4.8") {
      WowExpansion.MoP
    } else {
      throw new IllegalArgumentException(s"Version $version not supported!")
    }
  }
}

object ChatDirection extends Enumeration {
  type ChatDirection = Value
  val both, wow_to_discord, discord_to_wow = Value
}
